doc/dev/osd_internals/multi_object_write_transaction.rst

```diff
@@ -0,0 +1,117 @@
+==============================
+Multi-Object Write Transaction
+==============================
+
+Multi-object write transaction enables a group of write operations 
+on multi-objects complete atomically.
+
+Description of the Process
+--------------------------
+
+The high level process is,
+
+  1. Define the following class,
+
+     class *CEPH_RADOS_API MultiObjectWriteOperation*
+
+     {
+
+       std::map<std::string, ObjectWriteOperation*> slaves;
+
+       ObjectWriteOperation master;
+
+     };
+
+     based on this, client could send a group of ops on multi-objects 
+     through *MOSDOp* to the primary osd in charge of the *master* object 
+     which is explicitly specified by client, call this osd *MASTER*
+
+  #. *MASTER* receives the multi-object write operation request, performs 
+     dead lock check (detailed below), extracts *ObjectWriteOperations* 
+     by objects from *slaves*, and sends them to corresponding primary 
+     osds individually, calls those osds *SLAVE*, the transaction 
+     enters the *PREPARE* state
+
+  #. *SLAVE* receives *MOSDOp* from *MASTER*, note each op received by *SLAVE*
+     is a group of write operations on a single slave object. *SLAVE* 
+     performs dead lock check, does the normal *prepare_transaction()* stuff, 
+     if everything is ok, *SLAVE* will lock the object, and delay further 
+     read/write operations on this object until the transaction finishs. 
+     *SLAVE* sends a *QUERY* message to *MASTER*, and returns *MASTER* 
+     *EINPROGRESS*. The *QUERY* message is used for *MASTER* to notify *SLAVE* 
+     when to release transaction context, *MASTER* will not ack *QUERY* until 
+     itself has released transaction context
```

Comment:
I'm not sure that's a good tradeoff.  Deadlock detection by its nature detects a state which lasts for the duration of a single transaction.  Spending a whole round trip to detect a condition which could easily arise anyway in the mean time doesn't seem worthwhile.  It would make more sense to perform the deadlock detection atomically with the log step.  I think this is particularly important since a user of this mechanism could ensure that there is never a deadlock by construction.